File: MeatGrinder/Repositories/GoalRepository.cs
using System.Collections.Generic;
using System.Linq;
using MeatGrinder.Models;
using MeatGrinder.Services;

namespace MeatGrinder.Repositories
{
    public class GoalRepository
    {
        private readonly MeatGrinderEntities _db = new MeatGrinderEntities();

        public Goal GetByID(int id)
        {
           return _db.Goals.FirstOrDefault(m => m.ID == id);
        }
        public int GetChildTaskCount(int goalID)
        {
            return _db.Tasks.Count(m => m.GoalID == goalID && m.ParentTaskID == null);
        }
        public void SetGoalNotComplete(int goalId)
        {
            var goal = _db.Goals.FirstOrDefault(m => m.ID == goalId);

            if (goal != null && goal.IsComplete)
            {
                goal.IsComplete = false;
                _db.SaveChanges();
            }
        }
        public List<Goal> GetAllForUser()
        {
            int userID = CookieService.GetUserID();
            return _db.Goals.Where(m => m.UserID == userID).ToList();
        }
        public void Create(Goal goal)
        {
            goal.UserID = CookieService.GetUserID();
            goal.IsComplete = false;
            _db.Goals.Add(goal);
            _db.SaveChanges();
        }
        public void Delete(Goal goal)
        {
            var taskRepository = new TaskRepository();
            goal = _db.Goals.FirstOrDefault(m => m.ID == goal.ID);

            if (goal != null)
            {
                var childTasks = _db.Tasks.Where(m => m.GoalID == goal.ID);
                foreach (var childTask in childTasks)
                {
                    taskRepository.Delete(childTask);
                }
            }

            _db.Goals.Remove(goal);
            _db.SaveChanges();
        }
    }
}

File: MeatGrinder.Web/Repositories/GoalRepository.fs
namespace MeatGrinder.Web.Repositories
open System.Linq
//open Microsoft.FSharp.Linq.NullableOperators

open MeatGrinder.Orm

type GoalRepository (db:MeatGrinderEntities, getUserId:unit-> int option)=
    member x.GetById id =
        db.Goals.FirstOrDefault(fun m-> m.Id=id)
    member x.GetChildTaskCount goalId =
        db.Tasks.Count(fun m-> m.GoalId = goalId && m.ParentTaskID.HasValue=false)
    member x.SetGoalNotComplete goalId =
        let goal = db.Goals.FirstOrDefault(fun m->m.Id=goalId)
        if goal.Id>0 && goal.IsComplete then
            goal.IsComplete <- false
            db.SaveChanges() |> ignore
    member x.GetAllForUser () =
        let userId = getUserId().Value
        db.Goals.Where(fun m->m.UserId = userId) |> Seq.toList
    member x.Create (goal:Goal) =
        goal.UserId <- getUserId().Value
        goal.IsComplete <- false
        db.SaveChanges()
    member x.Delete deleteTasks goalId =
        let goal = db.Goals.FirstOrDefault(fun m->m.Id=goalId)
        if goal.Id>0 then
            deleteTasks <|| (db,db.Tasks.Where(fun m->m.GoalId = goalId))
            db.Goals.Remove goal |> ignore
            db.SaveChanges() |> ignore
    member x.UpdateChildTaskCounts (goals:Goal list) =
        let upd (g:Goal) = 
            g.ChildTaskCount <- Some(x.GetChildTaskCount(g.Id))
            if g.ChildTaskCount.Value > 0 then g.Description <- sprintf "%s(%i)" g.Description g.ChildTaskCount.Value
        goals |> Seq.iter upd
